### PR TITLE
Add viorng driver test scenerios

### PR DESCRIPTION
**Pull request by suqinhuang**
run random test before/during/after stop_contion/reboot/shutdown

Signed-off-by: Suqin Huang <shuang@redhat.com>

### CODE REVIEW

**Comment by suqinhuang** [comment]
id: 1231012 1231006 1230673

File: qemu/tests/cfg/viorng_in_use.cfg [hunk 1]
```diff
@@ -0,0 +1,43 @@
+- viorng_in_use:
+    type = driver_in_use
+    start_vm = yes
+    kill_vm_on_error = yes
+    check_guest_bsod = yes
+    login_timeout = 240
+    suppress_exception = True
+    driver_load_cmd = C:\devcon.exe enable @DRIVER_ID
+    driver_unload_cmd = C:\devcon.exe disable @DRIVER_ID
+    driver_id_cmd = C:\devcon.exe find * | find "VirtIO"
```
Comment (CongLi) [inline qemu/tests/cfg/viorng_in_use.cfg]: Where are the 3 lines above used?

File: qemu/tests/cfg/viorng_in_use.cfg [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    run_bgstress = rng_bat
+    target_process = D:\random_%PROCESSOR_ARCHITECTURE%.exe
+    rng_dll_register_cmd = "rundll32 C:\Windows\system32\viorngum.dll,RegisterProvider"
+    session_cmd_timeout = 240
+    read_rng_cmd  = "X:\random_%PROCESSOR_ARCHITECTURE%.exe"
+    enable_verifier_cmd = "verifier.exe /standard /driver viorng.sys"
```
Comment (CongLi) [inline qemu/tests/cfg/viorng_in_use.cfg]: In case "driver_in_use", driver verifier will be enabled, no need do it here.

Comment (suqinhuang) [inline qemu/tests/cfg/viorng_in_use.cfg]: this cmd is requested in rng_bat test, rem there is params to disable verifier?

Comment (CongLi) [inline qemu/tests/cfg/viorng_in_use.cfg]: I think it's better to add a flag here to check whether need enable verifier here.

Comment (suqinhuang) [inline qemu/tests/cfg/viorng_in_use.cfg]: https://github.com/avocado-framework/avocado-vt/pull/559

**Comment by suqinhuang** [comment]
https://github.com/avocado-framework/avocado-vt/pull/634

**Comment by pingl2016** [comment]
All during_bg_test cases and case viorng_in_use.with_shutdown.after_bg_test failed.
during_bg_test:
01:28:48 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/avocado_vt/test.py:422
01:28:48 ERROR| Traceback (most recent call last):
01:28:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/avocado_vt/test.py", line 197, in runTest
01:28:48 ERROR|     self._runTest()
01:28:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/avocado_vt/test.py", line 339, in _runTest
01:28:48 ERROR|     run_func(self, params, env)
01:28:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
01:28:48 ERROR|     return fn(*args, **kwargs)
01:28:48 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/driver_in_use.py", line 107, in run
01:28:48 ERROR|     stress_thread = run_bg_stress_test(bg_stress_test)
01:28:48 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/driver_in_use.py", line 63, in run_bg_stress_test
01:28:48 ERROR|     "alive!" % bg_stress_test)
01:28:48 ERROR| TestFail: Backgroud test rng_bat is not alive!
01:28:48 ERROR|
01:28:48 ERROR| FAIL 11-smp_2.4096m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_system_reset.during_bg_test -> TestFail: Backgroud test rng_bat is not alive!

viorng_in_use.with_shutdown.after_bg_test:
01:05:11 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/avocado_vt/test.py:422
01:05:11 ERROR| Traceback (most recent call last):
01:05:11 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/avocado_vt/test.py", line 197, in runTest
01:05:11 ERROR|     self._runTest()
01:05:11 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/avocado_vt/test.py", line 339, in _runTest
01:05:11 ERROR|     run_func(self, params, env)
01:05:11 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
01:05:11 ERROR|     return fn(*args, **kwargs)
01:05:11 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/driver_in_use.py", line 124, in run
01:05:11 ERROR|     utils_test.qemu.clear_win_driver_verifier(driver, vm, timeout)
01:05:11 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/utils_test/qemu.py", line 188, in clear_win_driver_verifier
01:05:11 ERROR|     session = vm.wait_for_login(timeout=timeout)
01:05:11 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/virt_vm.py", line 1032, in wait_for_login
01:05:11 ERROR|     self.verify_alive()
01:05:11 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/qemu_vm.py", line 177, in verify_alive
01:05:11 ERROR|     self.process.get_output())
01:05:11 ERROR| VMDeadError: VM is dead
01:05:11 ERROR|
01:05:11 ERROR| Traceback (most recent call last):

01:05:11 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/avocado_vt/test.py", line 436, in _run_avocado
    raise test_exception

01:05:11 ERROR| VMDeadError: VM is dead

**Comment by suqinhuang** [comment]
https://github.com/autotest/tp-qemu/pull/670

**Comment by CongLi** [comment]
15:39:01 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/rng_bat.py", line 68, in run
15:39:01 ERROR|     read_rng_cmd, enable_verifier_cmd, driver_verifier_cmd = session_cmds
15:39:01 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/rng_bat.py", line 67, in <genexpr>
15:39:01 ERROR|     for key in session_cmds_key)
15:39:01 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/rng_bat.py", line 48, in set_winutils_letter
15:39:01 ERROR|     return cmd.replace("X:", vol)
15:39:01 ERROR| AttributeError: 'NoneType' object has no attribute 'replace'

Can not find  "enable_verifier_cmd, driver_verifier_cmd " from the params.
@suqinhuang 
Could you help check it whether some updates are missed ?

Thanks.

**Comment by CongLi** [comment]
Same problem with @pingl2016 

As there is no "target_process" running in the guest, but case "driver_in_use" will check process running in the guest.

        if not utils_misc.wait_for(lambda: check_bg_running(session,
                                   target_process), 120, 0, 5):
            raise exceptions.TestFail("Backgroud test %s is not "
                                      "alive!" % bg_stress_test)

**Comment by suqinhuang** [comment]
PASS 1-smp_1.4096m.repeat1.run_test.Host_RHEL.m7.u2.rng_random.qcow2.virtio_blk.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_shutdown.before_bg_test
PASS 2-smp_1.4096m.repeat1.run_test.Host_RHEL.m7.u2.rng_random.qcow2.virtio_blk.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_shutdown.during_bg_test
PASS 3-smp_1.4096m.repeat1.run_test.Host_RHEL.m7.u2.rng_random.qcow2.virtio_blk.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_shutdown.after_bg_test

**Comment by CongLi** [comment]
With:
16:50:45 DEBUG|     shell_client = telnet
16:50:45 DEBUG|     shell_linesep = \r\n
16:50:45 DEBUG|     shell_port = 23

But still met error:
16:52:38 ERROR| TestFail: Unable to read random numbers fromguest: **** Error 0xc0000225 returned by BCryptOpenAlgorithmProvider

**Comment by suqinhuang** [comment]
is the driver installed, which rng_type did you use

**Comment by pingl2016** [comment]
\#/home/staf-kvm-devel/workspace/avocado-vt/scripts/scan_results.py 
CaseName                                 Status          StartTime       EndTime         Result     TimeElapsed
01-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_stop_continue.before_bg_test Finshed 2016-07-27 10:57:21 2016-07-27 10:59:39 PASS 138.504236
02-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_stop_continue.during_bg_test Finshed 2016-07-27 10:59:39 2016-07-27 11:02:01 PASS 141.653285
03-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_stop_continue.after_bg_test Finshed 2016-07-27 11:02:01 2016-07-27 11:04:35 PASS 154.295333
04-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_shutdown.before_bg_test Finshed 2016-07-27 11:04:35 2016-07-27 11:07:35 PASS 180.218744
05-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_shutdown.during_bg_test Finshed 2016-07-27 11:07:35 2016-07-27 11:08:31 PASS 55.291231
06-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_shutdown.after_bg_test Finshed 2016-07-27 11:08:31 2016-07-27 11:11:27 PASS 176.074368
07-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_reboot.before_bg_test Finshed 2016-07-27 11:11:27 2016-07-27 11:14:14 PASS 167.047529
08-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_reboot.during_bg_test Finshed 2016-07-27 11:14:14 2016-07-27 11:15:39 PASS 85.2773
09-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_reboot.after_bg_test Finshed 2016-07-27 11:15:39 2016-07-27 11:18:37 PASS 177.991524
10-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_system_reset.before_bg_test Finshed 2016-07-27 11:18:37 2016-07-27 11:21:42 PASS 184.465205
11-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_system_reset.during_bg_test Finshed 2016-07-27 11:21:42 2016-07-27 11:27:21 PASS 338.693846
12-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.viorng_in_use.with_system_reset.after_bg_test Finshed 2016-07-27 11:27:21 2016-07-27 11:30:45 PASS 203.901998
13-smp_4.8192m.repeat1.post_check.Host_RHEL.m7.u3.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.check_coredump Finshed 2016-07-27 11:30:45 2016-07-27 11:30:51 PASS 6.48477

Acked-by: Ping Li pingl@redhat.com

**Comment by CongLi** [comment]
For during with shutdown / system_reset cases, there is no random number generated and no random.exe from session log, but test pass.

Could you check it?

**Comment by suqinhuang** [comment]
@pingl2016 @CongLi 

follwoing is output of  viorng_in_use.with_shutdown.after_bg_test

and the random number will be generated when the process finish, sub test will be run before process finish, so no output for during process. I would provide the screendump if you need to check it.


2016-07-29 14:33:18: Generated random number is
2016-07-29 14:33:18: 0xdc, 0x16, 0x35, 0x7a, 0x8b, 0xfc, 0xad, 0xee, 0x57, 0x2d,
2016-07-29 14:33:18: 0x26, 0x0f, 0x7e, 0xf8, 0xba, 0x3f, 0x90, 0x2b, 0x21, 0xd9,
2016-07-29 14:33:18: 0x34, 0xce, 0xcf, 0xc1, 0xdf, 0x7f, 0x53, 0x24, 0x29, 0x5b,
2016-07-29 14:33:18: 0x38, 0xea, 0x18, 0xfe, 0xa6, 0xb5, 0x55, 0x64, 0xc0, 0x04,
2016-07-29 14:33:18: 0x1c, 0xbd, 0xc2, 0xf6, 0xaa, 0xa3, 0x7c, 0xe5, 0xc0, 0x7b,
2016-07-29 14:33:18: 0x67, 0x0a, 0xfd, 0xf1, 0xd3, 0x61, 0xbc, 0x24, 0xea, 0x1f,
2016-07-29 14:33:18: 0xec, 0x35, 0xe5, 0xae, 0xef, 0x32, 0xa3, 0x1b, 0x43, 0xf2,
2016-07-29 14:33:18: 0x0d, 0xe3, 0x68, 0xab, 0x90, 0x63, 0x91, 0x88, 0x1c, 0x97,
2016-07-29 14:33:18: 0xcf, 0xd3, 0x20, 0xcd, 0x46, 0x86, 0x43, 0x76, 0x0f, 0x35,
2016-07-29 14:33:18: 0xd3, 0xbe, 0x33, 0xba, 0x7a, 0xc1, 0xf6, 0x76, 0x84, 0xcf,
2016-07-29 14:33:18: 0x60, 0x5e, 0x4a, 0x07, 0x79, 0x7d, 0xb1, 0x73, 0x72, 0xd2,
2016-07-29 14:33:18: 0x58, 0xd5, 0xb4, 0x6c, 0x5f, 0x30, 0x5a, 0xd1, 0x3b, 0x87,
2016-07-29 14:33:18: 0x2e, 0xae, 0x6a, 0x52, 0x27, 0xa5, 0x38, 0x99,

**Comment by pingl2016** [comment]
It looks good to me now.

**Comment by CongLi** [comment]
There is no "random.exe" when execute "wmic process get name", but shutdown / system_reset is going.
I'm confused about it.

**Comment by suqinhuang** [comment]
this is the output if I print it:

Name                    
System Idle Process     
System                  
smss.exe                
csrss.exe               
wininit.exe             
csrss.exe               
winlogon.exe            
services.exe            
lsass.exe               
lsm.exe                 
svchost.exe             
svchost.exe             
svchost.exe             
svchost.exe             
svchost.exe             
svchost.exe             
svchost.exe             
spoolsv.exe             
svchost.exe             
taskhost.exe            
dwm.exe                 
explorer.exe            
tlntsvr.exe             
svchost.exe             
serialport.exe          
conhost.exe             
rss.exe                 
rss6.exe                
cmd.exe                 
conhost.exe             
SearchIndexer.exe       
tlntsess.exe            
conhost.exe             
cmd.exe                 
cmd.exe                 
WmiPrvSE.exe            
RANDOM_AMD64.EXE        --------------> here
SearchProtocolHost.exe  
SearchFilterHost.exe    
tlntsess.exe            
conhost.exe             
dllhost.exe             
cmd.exe                 
cmd.exe                 
WMIC.exe

**Comment by suqinhuang** [comment]
Hi @CongLi updated, and pass in your host

PASS 1-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.rng_random.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.viorng_in_use.with_shutdown.during_bg_test

**Comment by CongLi** [comment]
Acked-by: Cong Li <coli@redhat.com>

**Pull request merged**